colossus-metrics/src/main/scala/colossus/metrics/Collection.scala

```diff
@@ -117,29 +121,72 @@ class Collection(val config: CollectorConfig) {
         )
       }
     }
-    if (collectors.containsKey(address)) {
-      cast(collectors.get(address))
-    } else {
+    addContext(ns)
+    get(address).fold{
       val c = f(address, config)
-      collectors.putIfAbsent(address, c) match {
-        case null => c
-        case other => cast(other)
+      put(address, c) match {
+        case None => c
+        case Some(other) => cast(other)
       }
+    }{retrieved =>
+      cast(retrieved)
     }
   }
 
   def tick(interval: FiniteDuration): MetricMap = {
-    val keys = collectors.keys
     var build: MetricMap = Map()
-    while (keys.hasMoreElements) {
-      val key = keys.nextElement
-      Option(collectors.get(key)) foreach { c =>
-        build = build ++ c.tick(interval)
+    iteratorWithTags.foreach{ case (address, extraTags, c) =>
```

Comment:
I agree if it's not being used we can just remove the address building and have the iterator return a 2-tuple.